src/tap/src/tapcell.cpp

```diff
@@ -33,41 +33,1306 @@
+      const int urx = row_bb.xMax();
+      const int ury = row_bb.yMax();
+
+      odb::dbMaster* masterl;
+      odb::dbMaster* masterr;
+
+      if (do_corners) {
+        if (cur_row == top_row) {
+          masterl = pickCornerMaster(1,
+                                     row_ori,
+                                     cnrcap_nwin_master,
+                                     cnrcap_nwout_master,
+                                     endcap_master);
+          masterr = masterl;
+        } else if (cur_row == bottom_row) {
+          masterl = pickCornerMaster(-1,
+                                     row_ori,
+                                     cnrcap_nwin_master,
+                                     cnrcap_nwout_master,
+                                     endcap_master);
+          masterr = masterl;
+        } else {
+          auto rows_above = rows[cur_row + 1];
+          auto rows_below = rows[cur_row - 1];
+          masterl = pickCornerMaster(isXCorner(llx, rows_above, rows_below),
+                                     row_ori,
+                                     cnrcap_nwin_master,
+                                     cnrcap_nwout_master,
+                                     endcap_master);
+          masterr = pickCornerMaster(isXCorner(urx, rows_above, rows_below),
+                                     row_ori,
+                                     cnrcap_nwin_master,
+                                     cnrcap_nwout_master,
+                                     endcap_master);
+        }
+      } else {
+        masterl = endcap_master;
+        masterr = masterl;
+      }
+
+      if ((masterl->getWidth()) > (urx - llx)) {
+        continue;
+      }
+
+      buildCell(block, masterl, row_ori, llx, lly, endcap_prefix_);
+
+      int master_x = masterr->getWidth();
+      int master_y = masterr->getHeight();
+
+      int loc_2_x = urx - master_x;
+      int loc_2_y = ury - master_y;
+      if (llx == loc_2_x && lly == loc_2_y) {
+        logger_->warn(utl::TAP,
+                      9,
+                      "Row {} has enough space for only one endcap.",
+                      subrow->getName());
+        continue;
+      }
+
+      odb::dbOrientType right_ori = row_ori;
+      if (row_ori == odb::dbOrientType::MX) {
+        right_ori = odb::dbOrientType::R180;
+      } else if (row_ori == odb::dbOrientType::R0) {
+        right_ori = odb::dbOrientType::MY;
+      }
+      buildCell(block, masterr, right_ori, loc_2_x, loc_2_y, endcap_prefix_);
+    }
+  }
+
+  int endcap_count = phy_idx_ - start_phy_idx;
+  logger_->info(utl::TAP, 4, "Inserted {} endcaps.", endcap_count);
+  return endcap_count;
+}
+
+bool Tapcell::isXInRow(const int x, vector<odb::dbRow*>& subrow)
+{
+  for (odb::dbRow* row : subrow) {
+    odb::Rect row_bb;
+    row->getBBox(row_bb);
+    if (x >= row_bb.xMin() && x <= row_bb.xMax()) {
+      return true;
+    }
+  }
+  return false;
 }
 
+int Tapcell::isXCorner(const int x,
```

Comment:
'is' methods are usually boolean predicates.  What do the -1,0,1 return values mean?